File: tools/find_long.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# List dataset entries whose token length exceeds a threshold (diagnostic
# helper).
#
# Two counting modes (same as seqlen_stats.rb):
#   - with a model directory: the real tokenizer (tools/tokenizer.rb)
#   - without: the generator's conservative estimator
#
# Accepts both ShareGPT {"conversations": [...]} and MLX {"messages": [...]}.
#
# Usage:
#   ruby tools/find_long.rb dataset.jsonl [max_len] [limit]
#   ruby tools/find_long.rb dataset.jsonl <model_dir> [max_len] [limit]
#   (defaults: max_len = MAX_CONTEXT_TOKENS, limit = 20)

require "json"
require_relative "../build/build_dataset"
require_relative "tokenizer"

def messages_from(obj)
  if obj["messages"]
    obj["messages"]
  else
    obj["conversations"].map do |c|
      { "role" => c["from"] == "human" ? "user" : "assistant", "content" => c["value"] }
    end
  end
end

def entry_length(msgs, tokenizer)
  return tokenizer.count_messages(msgs) if tokenizer

  estimate_tokens(msgs.map { |m| m["content"].to_s }.join("\n")) + CHAT_TEMPLATE_OVERHEAD_TOKENS
end

# Returns [{ line:, len:, chars:, prompt: }] for the first `limit` entries of
# `lines` (any enumerable of JSONL lines) whose length exceeds `max_len`.
def find_over_limit(lines, max_len:, limit:, tokenizer: nil)
  results = []
  lines.each_with_index do |line, i|
    next if line.strip.empty?

    begin
      msgs = messages_from(JSON.parse(line))
    rescue JSON::ParserError => e
      warn "  [warn] line #{i + 1}: #{e.message}"
      next
    end

    len = entry_length(msgs, tokenizer)
    next unless len > max_len

    contents = msgs.map { |m| m["content"].to_s }
    chars = contents.sum(&:bytesize)
    prompt = contents.first.to_s.gsub("\n", " ")[0, 120]
    results << { line: i + 1, len: len, chars: chars, prompt: prompt }
    break if results.size >= limit
  end
  results
end

if __FILE__ == $PROGRAM_NAME
  PATH = ARGV[0]
  abort "usage: #{$PROGRAM_NAME} dataset.jsonl [model_dir] [max_len] [limit]" if PATH.nil? || !File.file?(PATH)

  REST = ARGV[1..]
  MODEL_DIR = REST.find { |a| File.directory?(a) }
  NUMS = REST.select { |a| a =~ /\A\d+\z/ }
  MAX_LEN = NUMS[0]&.to_i || MAX_CONTEXT_TOKENS
  LIMIT = NUMS[1]&.to_i || 20

  TOKENIZER = MODEL_DIR ? Tokenizer.new(MODEL_DIR) : nil

  results = find_over_limit(File.foreach(PATH, encoding: "UTF-8"),
                            max_len: MAX_LEN, limit: LIMIT, tokenizer: TOKENIZER)
  results.each do |r|
    label = TOKENIZER ? "#{r[:len]} tokens" : "~#{r[:len]} est tokens"
    puts format("line %d: %s, %d chars (c=%.2f) | %s",
                r[:line], label, r[:chars], r[:chars].to_f / r[:len], r[:prompt])
  end
  puts "(showing first #{results.size})"
end

File: tools/find_long.py
#!/usr/bin/env python3
"""List entries longer than a token threshold (diagnostic helper)."""
import json
import sys
from transformers import AutoTokenizer

path, model_dir = sys.argv[1], sys.argv[2]
max_len = int(sys.argv[3]) if len(sys.argv) > 3 else 2048
limit = int(sys.argv[4]) if len(sys.argv) > 4 else 20

tok = AutoTokenizer.from_pretrained(model_dir)
shown = 0
for i, line in enumerate(open(path, encoding="utf-8"), 1):
    obj = json.loads(line)
    msgs = obj.get("messages")
    if msgs is None:
        msgs = [
            {"role": "user" if c["from"] == "human" else "assistant", "content": c["value"]}
            for c in obj["conversations"]
        ]
    text = tok.apply_chat_template(msgs, tokenize=False)
    n = len(tok.encode(text))
    if n > max_len:
        shown += 1
        chars = sum(len(m["content"]) for m in msgs)
        prompt = msgs[0]["content"][:120].replace("\n", " ")
        print(f"line {i}: {n} tokens, {chars} chars (c={chars/n:.2f}) | {prompt}")
        if shown >= limit:
            break
print(f"(showing first {shown})")
